Gems/ONNX/Code/Source/Clients/ONNXSystemComponent.cpp

```diff
@@ -0,0 +1,300 @@
+        {
+            ONNXInterface::Register(this);
+        }
+
+        m_timingStats.SetName("MNIST Timing Statistics");
+        m_timingStats.SetHistogramBinCount(200);
+
+        if (ENABLE_CUDA)
+        {
+            m_timingStatsCuda.SetName("MNIST CUDA Timing Statistics");
+            m_timingStatsCuda.SetHistogramBinCount(200);
+        }
+
+        ImGui::ImGuiUpdateListenerBus::Handler::BusConnect();
+    }
+
+    ONNXSystemComponent::~ONNXSystemComponent()
+    {
+        ImGui::ImGuiUpdateListenerBus::Handler::BusDisconnect();
+
+        if (ONNXInterface::Get() == this)
+        {
+            ONNXInterface::Unregister(this);
+        }
+    }
+
+    Ort::Env* ONNXSystemComponent::GetEnv()
+    {
+        return m_env.get();
+    }
+
+    Ort::AllocatorWithDefaultOptions* ONNXSystemComponent::GetAllocator()
+    {
+        return m_allocator.get();
+    }
+
+    void OnnxLoggingFunction(
+        void*, OrtLoggingLevel, const char* category, const char* logid, const char* code_location, const char* message)
+    {
+        AZ_Printf("\nONNX", "%s %s %s %s", category, logid, code_location, message);
+    }
+
+    // The global environment and memory allocator are initialised with the system component, and are accessed via the EBus from within the
+    // model. m_precomputedTimingData and m_precomputedTimingDataCuda are structs holding the test inference statistics run before the
+    // editor starts up, and used by the ImGui dashboard.
+    void ONNXSystemComponent::Init()
+    {
+        void* ptr;
+        m_env = AZStd::make_unique<Ort::Env>(ORT_LOGGING_LEVEL_VERBOSE, "test_log", OnnxLoggingFunction, ptr);
+        m_allocator = AZStd::make_unique<Ort::AllocatorWithDefaultOptions>();
+        m_precomputedTimingData = AZStd::make_unique<PrecomputedTimingData>();
+        m_precomputedTimingDataCuda = AZStd::make_unique<PrecomputedTimingData>();
+    }
+
+    // Creates two mnist model instances that live in the system component and hooks them into the game tick. These are used for the
+    // realtime inferencing demo in the editor.
+    void ONNXSystemComponent::InitRuntimeMnistExamples()
+    {
+        ONNXRequestBus::Handler::BusConnect();
+        AZ::TickBus::Handler::BusConnect();
+
+        m_mnist = AZStd::make_unique<MNIST>();
+
+        AZStd::vector<float> input(m_mnist->m_imageSize);
+        m_mnist->m_input = input;
+
+        AZStd::vector<float> output(10);
+        m_mnist->m_output = output;
+
+        MNIST::InitSettings modelInitSettings;
+        modelInitSettings.m_inputShape = { 1, 1, 28, 28 };
+        modelInitSettings.m_outputShape = { 1, 10 };
+        modelInitSettings.m_modelName = "MNIST (Realtime)";
+
+        m_mnist->Load(modelInitSettings);
+
+        // For simplicity, the demo inferences the same test image on each tick.
+        std::string demoImage = std::string{ GEM_ASSETS_PATH } + std::string{ "/mnist_png/testing/0/10.png" };
+        m_mnist->LoadImage(demoImage.c_str());
```

Comment:
2 insights here.
1. std::string, should be used here, but AZ::IO::FixedMaxPath
2. AZ::IO::FileIOBase::ResolvePath should be used to get the gem root

```suggestion
        // For simplicity, the demo inferences the same test image on each tick.
        AZ::IO::FileIOBase fileIo = AZ::IO::FileIOBase::GetInstance();
        AZ::IO::FixedMaxPath demoImagePath;
        if (fileIoBase && fileIoBase->ResolvePath(demoImagePath, "@gemroot:ONNX@/Assets/mnist_png/testing/0/10.png")
        {
            m_mnist->LoadImage(demoImagePath.c_str());
        }
```

Reply:
Changed